### PR TITLE
fix StoreKit 2 decimal rounding

### PR DESCRIPTION
**Pull request by aboedo**
This is an ugly solution meant as starter code to illustrate a fix for #2207 

It seems like StoreKit 2's decimals for prices aren't rounded like you'd expect, leading to this issue which makes for ugly paywalls. 

This adds an extra step of rounding in a price-friendly way.

### CODE REVIEW

File: Sources/Purchasing/StoreKitAbstractions/SK2StoreProduct.swift [hunk 1]
```diff
@@ -58,7 +65,12 @@ internal struct SK2StoreProduct: StoreProductType {
         return offers?.first?["currencyCode"] as? String
     }
 
-    var price: Decimal { underlyingSK2Product.price }
+    var price: Decimal {
+        // See: https://github.com/RevenueCat/purchases-ios/issues/2207
+        // it seems like StoreKit 2 doesn't actually round decimals in a price-friendly fashion, so we do it
```
Comment (aboedo) [inline Sources/Purchasing/StoreKitAbstractions/SK2StoreProduct.swift:70]: See the words from the [StoreKit whisperer](https://stackoverflow.com/a/47445051/5074358)

File: Sources/Purchasing/StoreKitAbstractions/StoreProductDiscount.swift [hunk 2]
```diff
@@ -29,6 +29,13 @@ public typealias SK2ProductDiscount = StoreKit.Product.SubscriptionOffer
 @objc(RCStoreProductDiscount)
 public final class StoreProductDiscount: NSObject, StoreProductDiscountType {
 
+    private static let decimalRoundingBehavior = NSDecimalNumberHandler(roundingMode: .up,
+                                                                        scale: 2,
+                                                                        raiseOnExactness: false,
+                                                                        raiseOnOverflow: false,
+                                                                        raiseOnUnderflow: false,
+                                                                        raiseOnDivideByZero: false)
```
Comment (aboedo) [inline Sources/Purchasing/StoreKitAbstractions/StoreProductDiscount.swift:37]: yeah, not exactly clean to have this duplicated for both classes, I was going for "quick PR to show off a possible solution"

File: Sources/Purchasing/StoreKitAbstractions/StoreProductDiscount.swift [hunk 3]
```diff
@@ -70,7 +77,13 @@ public final class StoreProductDiscount: NSObject, StoreProductDiscountType {
     @objc public var offerIdentifier: String? { self.discount.offerIdentifier }
     @objc public var currencyCode: String? { self.discount.currencyCode }
     // See also `priceDecimalNumber` for Objective-C
-    public var price: Decimal { self.discount.price }
+    public var price: Decimal {
+        // See: https://github.com/RevenueCat/purchases-ios/issues/2207
+        // it seems like StoreKit 2 doesn't actually round decimals in a price-friendly fashion, so we do it
+        let priceNSDecimal = (self.discount.price as NSDecimalNumber)
```
Comment (aboedo) [inline Sources/Purchasing/StoreKitAbstractions/StoreProductDiscount.swift:83]: ended up going back and forth to objc here, I bet there's a nicer solution

File: Sources/Purchasing/StoreKitAbstractions/StoreProductDiscount.swift [hunk 4]
```diff
@@ -70,7 +77,13 @@ public final class StoreProductDiscount: NSObject, StoreProductDiscountType {
...
+        return priceNSDecimal.rounding(accordingToBehavior: Self.decimalRoundingBehavior) as Decimal
```
Comment (aboedo) [inline Sources/Purchasing/StoreKitAbstractions/StoreProductDiscount.swift:84]: for writing tests for this, we could try to reproduce using SKTest. There's a chance that SKTest doesn't actually reproduce, though, and that we end up having to just mock with unit tests.

Which... we can't do directly because SK2 so we'd have to move this logic to some other class or something and test that one instead, then inject it here and test that we're calling the class.

**Comment by aboedo** [comment]
@NachoSoto would you mind taking it from here?

File: Sources/Purchasing/StoreKitAbstractions/StoreProductDiscount.swift (hunk 4 above)
Comment (aboedo) [reply Sources/Purchasing/StoreKitAbstractions/StoreProductDiscount.swift:84]: here's code that I used to test that the approach works: 

```swift
let price = 114.98999999999997952

print(price)

let priceDecimal = NSDecimalNumber(floatLiteral: price)

print(priceDecimal)

let behavior = NSDecimalNumberHandler(roundingMode: .up, scale: 2, raiseOnExactness: false, raiseOnOverflow: false, raiseOnUnderflow: false, raiseOnDivideByZero: false)

let roundedDecimal = priceDecimal.rounding(accordingToBehavior: behavior)

let swiftDecimal = Decimal(floatLiteral: price)

print(swiftDecimal)

let roundedSwift = (swiftDecimal as NSDecimalNumber).rounding(accordingToBehavior: behavior)

let finalDecimal = roundedSwift as Decimal

print(finalDecimal)

```

Comment (NachoSoto) [reply Sources/Purchasing/StoreKitAbstractions/StoreProductDiscount.swift:84]: Based on this I think maybe the best solution is to extract the rounding so we can test it like this, instead of bothering with StoreKit.

File: Sources/Purchasing/StoreKitAbstractions/StoreProductDiscount.swift (hunk 3 above)
Comment (NachoSoto) [reply Sources/Purchasing/StoreKitAbstractions/StoreProductDiscount.swift:83]: Hmmm yeah let me look at these.

**Comment by NachoSoto** [comment]
Sure!

**Review by NachoSoto — CHANGES_REQUESTED** [review]
The more I think about this, I don't think we should modify the actual value. `Decimal` is essentially a `Float` with a fancy type.
The fact that printing it shows `114.98999999999997952` is a typical consequence of binary float number representation.
`Decimal.description` doesn't say anything about formatting.

This is why we format it explicitly using `NSDecimalNumber` [when posting it to the backend](https://github.com/RevenueCat/purchases-ios/blob/4.16.0/Sources/Purchasing/ProductRequestData.swift#L87-L88).

```swift
NSDecimalNumber(floatLiteral: 114.98999999999997952) == NSDecimalNumber(floatLiteral: 114.99) // true
```

So in other words, we shouldn't modify the underlying `float` value. What if one of the locales uses a precision different than 2 decimals? We'd be incorrectly rounding the value.

**Comment by aboedo** [comment]
> So in other words, we shouldn't modify the underlying float value. What if one of the locales uses a precision different than 2 decimals? We'd be incorrectly rounding the value.

Ooff, yeah, I forgot about that, good catch. There are a few currencies that use 3 decimals

**Comment by NachoSoto** [comment]
I would say the issue is a won't-fix. Doing `p price` is relying on LLDB to format a `Float`, which is not how one would format a price in the UI, and we do offer different ways to format prices already.
